### PR TITLE
fix: complex arguments to builtin functions

### LINKED ISSUE #3164: Multiple evaluation of argument leads to non-unique symbol error in some built-in fuctions
### Version Information

* vyper Version (output of `vyper --version`): 0.3.8+commit.2d18e60b
* OS: Linux (Ubuntu 22.04.1 LTS)
* Python Version (output of `python --version`): 3.10.6

### What's your issue about?

The argument of the [ceil](https://vyper.readthedocs.io/en/stable/built-in-functions.html#ceil) built-in is evaluated multiple times. This leads to duplicate symbol errors if the argument is a function call. See the [build_IR](https://github.com/vyperlang/vyper/blob/master/vyper/builtins/functions.py#L183) function in Ceil in vyper/builtins/functions.py for the code responsible.

Example of code that crashes:
```vyper
@external
def foo() -> (int256):
    a: Foo = Foo(msg.sender)

    return ceil(a.bar())

interface Foo:
    def bar() -> decimal: nonpayable
```
`AssertionError: non-unique symbols {'a.bar()2'}`

Other built-in functions where this occurs:
- `floor` where the argument `value` is evaluated multiple times
- `as_wei_value` with argument `value` 
- `uint256_addmod` with argument `c`
- `uint256_mulmod` with argument `c`

### How can it be fixed?

Use something like [cache_when_complex](https://github.com/vyperlang/vyper/blob/master/vyper/codegen/ir_node.py#L388) to evaluate the argument once, and store it for re-use.

### PR DESCRIPTION
**Pull request by tserg**
### What I did

Fix #3164, and also `ecadd` and `ecmul`.

### How I did it

Cache the argument.

### How to verify it

See tests.

### Commit message

```
prior to this commit, some builtin functions including ceil, would panic
if their arguments were function calls (or otherwise determined to be
complex expressions by `is_complex_ir`). this commit fixes the relevant
builtin functions by using `cache_when_complex` where appropriate.
```

### Description for the changelog

Fix codegen for function calls as argument in builtin functions

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Review by fubuloubu — APPROVED** [review]

**Comment by charles-cooper** [comment]
also, since the test functionality is repeated (here, and also in other places but we don't need to worry about the other places just now), it could be good to have some scaffolding / utilities which make this easier. i'm thinking like, a common contract which is intended to be called for its side effects, and a utility function which checks that the side effect was invoked exactly one time.

**Comment by tserg** [comment]
I have added two fixtures `side_effects_contract` and `assert_side_effect_invoked_once` to standardise the test for side effects. What do you think?

**Comment by tserg** [comment]
I am not sure why this call to `foo()` reverts if this contract was in the test suite. As long as there is more than one external call in `uint256_addmod` regardless of the argument position, the transaction reverts even if the gas is set to a high value. If tested in boa, `foo()` works fine.
```
@external
def foo(addr: address) -> uint256:
    f: Foo = Foo(addr)
    return uint256_addmod(f.a(), f.b(), f.c())

interface Foo:
    def a() -> uint256: payable
    def b() -> uint256: payable
    def c() -> uint256: payable
```

File: tests/conftest.py [hunk 1]
```diff
@@ -201,3 +201,54 @@ def _f(_addr, _salt, _initcode):
...
+        Takes in a list of tuple of (
+            function_name: Function to be called by the test contract
+            return_type: Return type of the function
+            return_value: Return value of the function
+        ) and generates a Vyper contract for testing side effects.
+        """
+        code = ""
+
+        for i in inputs:
+            assert len(i) == 3
```
Comment (charles-cooper) [inline tests/conftest.py:219]: i think the assert is actually not needed because otherwise the unpacking will fail on the next line

**Review by charles-cooper — APPROVED** [review]
cleaned up the side effects checker API a little in https://github.com/vyperlang/vyper/pull/3167/commits/532348175190687e651513625da1a7074251580d

**Pull request merged**